hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/marker/MarkerDirState.java

```diff
@@ -0,0 +1,393 @@
+      markerCreationFutures.add(future);
+    }
+  }
+
+  /**
+   * @return the next file index to use in a round-robin fashion,
+   * or -1 if no file is available.
+   */
+  public int getNextFileIndexToUse() {
+    int fileIndex = -1;
+    synchronized (threadUseStatus) {
+      int nextIndex = (lastFileIndexUsed + 1) % threadUseStatus.size();
+      if (!threadUseStatus.get(nextIndex)) {
+        fileIndex = nextIndex;
+        threadUseStatus.set(nextIndex, true);
+      } else {
+        for (int i = 1; i < threadUseStatus.size(); i++) {
+          int index = (lastFileIndexUsed + 1 + i) % threadUseStatus.size();
+          if (!threadUseStatus.get(index)) {
+            fileIndex = index;
+            threadUseStatus.set(index, true);
+            break;
+          }
+        }
+      }
+      if (fileIndex >= 0) {
+        lastFileIndexUsed = fileIndex;
+      }
+    }
+    return fileIndex;
+  }
+
+  /**
+   * Marks the file as available to use again.
+   *
+   * @param fileIndex file index
+   */
+  public void markFileAsAvailable(int fileIndex) {
+    synchronized (threadUseStatus) {
+      threadUseStatus.set(fileIndex, false);
+    }
+  }
+
+  /**
+   * @return  futures of pending marker creation requests and removes them from the list.
+   */
+  public List<MarkerCreationCompletableFuture> fetchPendingMarkerCreationRequests() {
+    if (markerCreationFutures.isEmpty()) {
+      return new ArrayList<>();
+    }
+    maybeSyncOnFirstRequest();
+
+    List<MarkerCreationCompletableFuture> pendingFutures;
+    synchronized (markerCreationFutures) {
+      pendingFutures = new ArrayList<>(markerCreationFutures);
+      markerCreationFutures.clear();
+    }
+    return pendingFutures;
+  }
+
+  /**
+   * Processes pending marker creation requests.
+   *
+   * @param pendingMarkerCreationFutures futures of pending marker creation requests
+   * @param fileIndex file index to use to write markers
+   */
+  public void processMarkerCreationRequests(
+      final List<MarkerCreationCompletableFuture> pendingMarkerCreationFutures, int fileIndex) {
+    if (pendingMarkerCreationFutures.isEmpty()) {
+      markFileAsAvailable(fileIndex);
+      return;
+    }
+
+    LOG.debug("timeMs=" + System.currentTimeMillis() + " markerDirPath=" + markerDirPath
+        + " numRequests=" + pendingMarkerCreationFutures.size() + " fileIndex=" + fileIndex);
+
+    synchronized (markerCreationProcessingLock) {
+      for (MarkerCreationCompletableFuture future : pendingMarkerCreationFutures) {
+        String markerName = future.getMarkerName();
+        boolean exists = allMarkers.contains(markerName);
+        if (!exists) {
+          allMarkers.add(markerName);
+          StringBuilder stringBuilder = fileMarkersMap.computeIfAbsent(fileIndex, k -> new StringBuilder(16384));
+          stringBuilder.append(markerName);
+          stringBuilder.append('\n');
+        }
+        future.setResult(!exists);
+      }
+    }
+    flushMarkersToFile(fileIndex);
```

Comment:
if we fail here due to s3 error lets say. then the in-memory state is not consistent? for future requests? may be it does not matter tht there are marker entries that have no file associated (this case can always happen if the executor dies right after the marker request returns and before the actual data file is created)

Reply:
Also, since we time out the requests in the batch if S3 write IOException happens, the executors retry with new marker creations requests.  In such a case, for rollback, the timeline server returns extra invalid markers which are not useful anyway, not affecting the functionality.